keras_nlp/layers/random_replacement.py

```diff
@@ -0,0 +1,379 @@
+                "Input must either be rank 1 or rank 2. Received input with "
+                f"rank={inputs.shape.rank}"
+            )
+        elif inputs.shape.rank == 1:
+            input_is_1d = True
+            # Add a new axis at the beginning.
+            inputs = tf.expand_dims(inputs, axis=0)
+        if isinstance(inputs, tf.Tensor):
+            # Convert to ragged tensor.
+            inputs = tf.RaggedTensor.from_tensor(inputs)
+
+        skip_masks = None
+        if self.skip_list:
+            skip_masks = self.StaticHashTable.lookup(inputs.flat_values)
+        elif self.skip_fn:
+            skip_masks = tf.map_fn(
+                self.skip_fn, inputs.flat_values, fn_output_signature=tf.bool
+            )
+        elif self.skip_py_fn:
+
+            def string_fn(token):
+                return self.skip_py_fn(token.numpy().decode("utf-8"))
+
+            def int_fn(token):
+                return self.skip_py_fn(token.numpy())
+
+            py_fn = string_fn if inputs.dtype == tf.string else int_fn
+
+            skip_masks = tf.map_fn(
+                lambda x: tf.py_function(py_fn, [x], tf.bool),
+                inputs.flat_values,
+                fn_output_signature=tf.bool,
+            )
+
+        positions = tf.ragged.range(inputs.row_lengths())
+
+        if skip_masks is not None:
+            skip_masks = tf.logical_not(skip_masks)
+            skip_masks.set_shape([None])
+            positions = tf.ragged.boolean_mask(
+                positions, inputs.with_flat_values(skip_masks)
+            )
+        # Figure out how many we are going to select.
+        token_counts = tf.cast(positions.row_lengths(), "float32")
+        num_to_select = tf.random.stateless_binomial(
+            shape=tf.shape(token_counts),
+            seed=self._generator.make_seeds()[:, 0],
+            counts=token_counts,
+            probs=self.rate,
+        )
+        if self.max_replacements is not None:
+            num_to_select = tf.math.minimum(
+                num_to_select, self.max_replacements
+            )
+        num_to_select = tf.math.minimum(
+            num_to_select, tf.cast(positions.row_lengths(), tf.int32)
+        )
+        num_to_select = tf.cast(num_to_select, tf.int64)
+
+        def _replace(x):
+            inputs, num_to_select, positions = x
+            for _ in range(num_to_select):
+                # Choose a Random Index
+                index = positions[
+                    tf.random.stateless_uniform(
+                        shape=[],
+                        minval=0,
+                        maxval=tf.size(positions),
+                        dtype=tf.int32,
+                        seed=self._generator.make_seeds()[:, 0],
+                    )
+                ]
+                synonym = inputs[index]
+
+                if self.replacement_fn is not None:
+                    synonym = self.replacement_fn(synonym)
+                    inputs = tf.tensor_scatter_nd_update(
```

Comment:
I'm skeptical we really need a `map_fn` with a loop inside it, with a scatter inside it. That sounds like it will be inefficient.

If we follow the flow we have in random deletion, we will have a complete `mask` containing only the indices we want to run a deletion on right? Then we could do something like run a `tf.map_fn` over the pair of `(inputs.flat_values, mask.flat_values)` and early return if mask == false, otherwise we lookup a replacement.

There might be other ways to do this, but overall we should:
 - avoid nested maps/loops
 - avoid scatters inside a map/loop